File: application/backend/src/plaid_service.py
# application/backend/src/plaid_service.py

import plaid
from plaid.api import plaid_api
from plaid.configuration import Environment
from plaid.model.link_token_create_request import LinkTokenCreateRequest
from plaid.model.item_public_token_exchange_request import (
    ItemPublicTokenExchangeRequest,
)
from plaid.model.country_code import CountryCode
from plaid.model.products import Products
from application.backend.src.config import BackendConfig as Config


def initialize_plaid_client():
    """
    Initialize the Plaid client.
    :return: Plaid client object
    """
    Config.validate()
    client_id = Config.PLAID_CLIENT_ID
    secret = Config.PLAID_SECRET_KEY
    environment = Config.PLAID_ENV

    environment_mapping = {
        "sandbox": Environment.Sandbox,
        "production": Environment.Production,
    }

    if environment == "production":
        print("WARNING: Plaid client is in production mode")

    configuration = plaid.configuration.Configuration(
        host=environment_mapping[environment],
        api_key={"clientId": client_id, "secret": secret},
    )

    api_client = plaid_api.ApiClient(configuration)
    return plaid_api.PlaidApi(api_client)


def create_link_token(client: object, user_id, country_codes, products):
    """
    Create a link token for the user.
    :param client: Plaid client object
    :param user_id: User ID
    :param country_codes: Country codes
    :param products: Products
    :return: Link token
    """
    try:
        request = LinkTokenCreateRequest(
            user={"client_user_id": user_id},
            products=[Products(products)],
            client_name="Personal Finance App",
            country_codes=[CountryCode(country_codes)],
            language="en",
        )
        response = client.link_token_create(request)
        return response["link_token"]

    except plaid.exceptions.ApiException as e:
        print(f"Exception when calling PlaidApi: {e}")
        return None


def exchange_public_token(client: object, public_token):
    """
    Exchange a public token for an access token.
    :param client: Plaid client object
    :param public_token: Public token
    :return: Access token
    """
    try:
        request = ItemPublicTokenExchangeRequest(public_token=public_token)
        response = client.item_public_token_exchange(request)
        return response["access_token"]

    except plaid.exceptions.ApiException as e:
        print(f"Exception when calling PlaidApi: {e}")
        return None

File: apps/api/src/plaid/plaid.service.ts
import {
  BadRequestException,
  Inject,
  Injectable,
  ServiceUnavailableException,
} from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import {
  Configuration,
  CountryCode,
  PlaidApi,
  PlaidEnvironments,
  Products,
} from "plaid";

import { PrismaService } from "../common/prisma.service.js";
import { TokenEncryptionService } from "./token-encryption.service.js";

@Injectable()
export class PlaidService {
  private readonly plaidClient: PlaidApi;
  private readonly plaidClientId: string | undefined;
  private readonly plaidSecret: string | undefined;
  private readonly products: Products[];
  private readonly countryCodes: CountryCode[];
  private readonly plaidRedirectUri: string | null;

  constructor(
    @Inject(ConfigService) private readonly configService: ConfigService,
    @Inject(PrismaService) private readonly prisma: PrismaService,
    @Inject(TokenEncryptionService)
    private readonly tokenEncryptionService: TokenEncryptionService,
  ) {
    this.plaidClientId = this.configService.get<string>("PLAID_CLIENT_ID");
    const plaidEnv = this.configService.get<string>("PLAID_ENV", "sandbox");

    const legacySecret = this.configService.get<string>("PLAID_SECRET");
    const sandboxSecret = this.configService.get<string>(
      "PLAID_SANDBOX_SECRET",
    );
    const developmentSecret = this.configService.get<string>(
      "PLAID_DEVELOPMENT_SECRET",
    );
    const productionSecret = this.configService.get<string>(
      "PLAID_PRODUCTION_SECRET",
    );

    this.plaidSecret =
      (plaidEnv === "production" ? productionSecret : undefined) ??
      (plaidEnv === "development" ? developmentSecret : undefined) ??
      (plaidEnv === "sandbox" ? sandboxSecret : undefined) ??
      legacySecret;

    this.products = this.configService
      .get<string>("PLAID_PRODUCTS", "transactions")
      .split(",")
      .map((value) => value.trim())
      .filter(Boolean) as Products[];

    this.countryCodes = this.configService
      .get<string>("PLAID_COUNTRY_CODES", "US")
      .split(",")
      .map((value) => value.trim())
      .filter(Boolean) as CountryCode[];

    this.plaidRedirectUri =
      this.configService.get<string>("PLAID_REDIRECT_URI") ?? null;

    this.plaidClient = new PlaidApi(
      new Configuration({
        basePath: PlaidEnvironments[plaidEnv as keyof typeof PlaidEnvironments],
        baseOptions: {
          headers: {
            "PLAID-CLIENT-ID": this.plaidClientId ?? "",
            "PLAID-SECRET": this.plaidSecret ?? "",
          },
        },
      }),
    );
  }

  getPlaidClient(): PlaidApi {
    this.ensurePlaidConfigured();
    return this.plaidClient;
  }

  async createLinkToken(userId: string): Promise<{ linkToken: string }> {
    this.ensurePlaidConfigured();
    await this.ensureUserExists(userId);

    const request = {
      user: {
        client_user_id: userId,
      },
      client_name: "Personal Finances",
      products: this.products,
      country_codes: this.countryCodes,
      language: "en",
      ...(this.plaidRedirectUri ? { redirect_uri: this.plaidRedirectUri } : {}),
    };

    const response = await this.plaidClient.linkTokenCreate(request);

    return {
      linkToken: response.data.link_token,
    };
  }

  async exchangePublicToken(
    userId: string,
    publicToken: string,
  ): Promise<{ userId: string; linked: boolean }> {
    this.ensurePlaidConfigured();
    await this.ensureUserExists(userId);

    const exchangeResponse = await this.plaidClient.itemPublicTokenExchange({
      public_token: publicToken,
    });

    await this.prisma.plaidItem.upsert({
      where: {
        plaidItemId: exchangeResponse.data.item_id,
      },
      create: {
        userId,
        plaidItemId: exchangeResponse.data.item_id,
        accessToken: this.tokenEncryptionService.encryptAccessToken(
          exchangeResponse.data.access_token,
        ),
      },
      update: {
        userId,
        accessToken: this.tokenEncryptionService.encryptAccessToken(
          exchangeResponse.data.access_token,
        ),
      },
    });

    return {
      userId,
      linked: true,
    };
  }

  async reencryptLegacyAccessTokens(): Promise<{
    scanned: number;
    updated: number;
  }> {
    return this.tokenEncryptionService.reencryptLegacyAccessTokens();
  }

  private ensurePlaidConfigured(): void {
    if (!this.plaidClientId || !this.plaidSecret) {
      throw new ServiceUnavailableException("Missing Plaid credentials");
    }
  }

  private async ensureUserExists(userId: string): Promise<void> {
    const user = await this.prisma.user.findUnique({ where: { id: userId } });
    if (!user) {
      throw new BadRequestException("Unknown user");
    }
  }
}
